crates/solidity/inputs/language/src/bindings.rs

```diff
@@ -1,55 +1,141 @@
+    writeln!(buffer, "use semver::Version;")?;
+
+    writeln!(buffer, "#[allow(clippy::too_many_lines)]")?;
+    writeln!(
+        buffer,
+        "\npub fn define_built_ins<KT: KindTypes + 'static>(builder: &mut FileGraphBuilder<'_, KT>, scope: &mut ScopeGraphBuilder, version: &Version) {{",
+    )?;
+    let versions = language.collect_built_ins_versions();
+    let versions = versions.iter().collect::<Vec<_>>();
+    for vs in versions.windows(2) {
+        let version = vs[0];
+        let next_version = vs[1];
+        writeln!(
+            buffer,
+            "if *version < Version::new({major}, {minor}, {patch}) {{",
+            major = next_version.major,
+            minor = next_version.minor,
+            patch = next_version.patch
+        )?;
+        writeln!(buffer, "    // {version}")?;
+
+        render_contexts_for_version(&mut buffer, &language.built_ins, version)?;
+
+        write!(buffer, "  }} else ")?;
+    }
+    writeln!(buffer, "{{")?;
+    let last_version = versions.last().unwrap();
+    writeln!(buffer, "  // {last_version}")?;
+    render_contexts_for_version(&mut buffer, &language.built_ins, last_version)?;
+    writeln!(buffer, "  }}")?;
+    writeln!(buffer, "}}")?;
+
+    Ok(buffer)
 }
 
-fn try_render_built_ins(built_ins: &[BuiltIn]) -> Result<String, Error> {
-    let mut buffer = String::new();
-    // __SLANG_SOLIDITY_BUILT_INS_CONTRACT_NAME__ keep in sync with binding rules
-    writeln!(buffer, "contract $BuiltIns$ {{")?;
-    for item in built_ins {
-        match item {
-            BuiltIn::BuiltInFunction { item } => {
-                writeln!(
-                    buffer,
-                    "  function {name}({parameters}) public{return_type};",
-                    name = item.name,
-                    parameters = item.parameters.join(", "),
-                    return_type = item
-                        .return_type
-                        .as_ref()
-                        .map(|return_type| format!(" returns ({return_type})"))
-                        .unwrap_or_default(),
-                )?;
-            }
-            BuiltIn::BuiltInType { item } => {
-                writeln!(buffer, "  struct {name} {{", name = item.name)?;
-                for field in &item.fields {
-                    writeln!(buffer, "    {field_def};", field_def = field.definition)?;
+fn tr(input: &str) -> String {
+    match input {
+        "int" => "int256".into(),
+        "uint" => "uint256".into(),
+        _ => input.replace('$', "%"),
+    }
+}
+
+fn render_contexts_for_version(
+    buffer: &mut String,
+    built_in_contexts: &[BuiltInContext],
+    version: &Version,
+) -> Result<(), Error> {
+    for context in built_in_contexts {
+        writeln!(buffer, "    // {context}", context = context.name)?;
+        writeln!(buffer, "    {{")?;
+        // __SLANG_YUL_BUILT_INS_CONTEXT_NAME__ keep in sync with language definition
+        if context.name == "$YulBuiltIns$" {
```

Comment:
That means that *ALL* Solidity built-ins are accessible from Yul. Right?

Reply:
No, not quite. Because reference nodes created from Yul identifiers are still guarded by a `@yul` symbol before they can reach the built-ins scope.

Reply:
Do we have tests that cover these cases?

- that we don't bind a Solidity built-in in a Yul context.
- that we don't bind a Yul built-in in a Solidity context.

Reply:
Added tests for both cases. Solidity built-ins *were* being bound in Yul contexts, so I updated the binding rules to fix that.